rpcs3/Emu/Memory/vm.cpp

```diff
@@ -1,24 +1,324 @@
+		__noinline void lock()
+		{
+			NamedThreadBase* owner = GetCurrentNamedThread();
+			NamedThreadBase* old = nullptr;
+
+			while (!m_owner.compare_exchange_strong(old, owner))
+			{
+				std::unique_lock<std::mutex> cv_lock(m_cv_mutex);
+
+				m_cv.wait_for(cv_lock, std::chrono::milliseconds(1));
+
+				if (old == owner)
+				{
+					throw __FUNCTION__;
+				}
+
+				old = nullptr;
+			}
+
+			do_notify = true;
+
+			test++;
+			assert(test == 1);
+		}
+
+		__noinline void unlock()
+		{
+			assert(test == 1);
+			test--;
+
+			NamedThreadBase* owner = GetCurrentNamedThread();
+
+			if (!m_owner.compare_exchange_strong(owner, nullptr))
+			{
+				throw __FUNCTION__;
+			}
+
+			if (do_notify)
+			{
+				m_cv.notify_one();
+			}
+		}
+
+	};
+
+	std::function<void()> g_reservation_cb = nullptr;
+	NamedThreadBase* g_reservation_owner = nullptr;
+
+	u32 g_reservation_addr = 0;
+
+	reservation_mutex_t g_reservation_mutex;
+
+	void _reservation_set(u32 addr)
+	{
+		//const auto stamp0 = get_time();
+
+#ifdef _WIN32
+		DWORD old;
+		if (!VirtualProtect(vm::get_ptr(addr & ~0xfff), 4096, PAGE_READONLY, &old))
+#else
+		if (mprotect(vm::get_ptr(addr & ~0xfff), 4096, PROT_READ))
+#endif
+		{
+			throw fmt::format("vm::_reservation_set() failed (addr=0x%x)", addr);
+		}
+
+		//LOG_NOTICE(MEMORY, "VirtualProtect: %f us", (get_time() - stamp0) / 80.f);
+	}
+
+	bool _reservation_break(u32 addr)
+	{
+		if (g_reservation_addr >> 12 == addr >> 12)
+		{
+			//const auto stamp0 = get_time();
+
+#ifdef _WIN32
+			if (!VirtualAlloc(vm::get_ptr(addr & ~0xfff), 4096, MEM_COMMIT, PAGE_READWRITE))
+#else
+			if (mprotect(vm::get_ptr(addr & ~0xfff), 4096, PROT_READ | PROT_WRITE))
+#endif
+			{
+				throw fmt::format("vm::_reservation_break() failed (addr=0x%x)", addr);
+			}
+
+			//LOG_NOTICE(MEMORY, "VirtualAlloc: %f us", (get_time() - stamp0) / 80.f);
+
+			if (g_reservation_cb)
+			{
+				g_reservation_cb();
+				g_reservation_cb = nullptr;
+			}
+
+			g_reservation_owner = nullptr;
+			g_reservation_addr = 0;
+
+			return true;
+		}
+
+		return false;
+	}
+
+	bool reservation_break(u32 addr)
+	{
+		std::lock_guard<reservation_mutex_t> lock(g_reservation_mutex);
+
+		return _reservation_break(addr);
+	}
+
+	bool reservation_acquire(void* data, u32 addr, u32 size, const std::function<void()>& callback)
+	{
+		const auto stamp0 = get_time();
+
+		bool broken = false;
+
+		assert(size == 1 || size == 2 || size == 4 || size == 8 || size == 128);
+		assert((addr + size & ~0xfff) == (addr & ~0xfff));
+
+		{
+			std::lock_guard<reservation_mutex_t> lock(g_reservation_mutex);
+
+			// silent unlocking to prevent priority boost for threads going to break reservation
+			g_reservation_mutex.do_notify = false;
+
+			// break previous reservation
+			if (g_reservation_addr)
+			{
+				broken = _reservation_break(g_reservation_addr);
+			}
+
+			// change memory protection to read-only
+			_reservation_set(addr);
+
+			// set additional information
+			g_reservation_addr = addr;
```

Comment:
My understanding is that each SPU and PPU in the Cell can make one reservation at a time and as long as the reservation addresses do not conflict, each reservation will remain valid. It seems this implementation allows for only one reservation per Cell.